### PR TITLE
STARCH-887 Split Crypto API

### PR DESCRIPTION
**Pull request by mwfyffeiu**

### CODE REVIEW

File: crypt/impl/src/test/java/iu/crypt/PemEncodedTest.java [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package iu.crypt;
+
+@SuppressWarnings("javadoc")
+public class PemEncodedTest {
```
Comment (james-anderson-iu) [inline crypt/impl/src/test/java/iu/crypt/PemEncodedTest.java:35]: Is this test class still needed? It looks like PemEncoded and PemEncodedTest were moved to the crypt api module.

**Review by james-anderson-iu — COMMENTED** [review]
Do we have any cases when converting between types with `from` and `to`, such as in `CryptJsonAdapters`, where it would be necessary to test both `from(to())` and `to(from())`?

File: crypt/impl/src/test/java/iu/crypt/PemEncodedTest.java (hunk 1 above)
Comment (mwfyffeiu) [reply crypt/impl/src/test/java/iu/crypt/PemEncodedTest.java]: It is not needed and was removed by the latest commit

File: crypt/impl/src/main/java/iu/crypt/CompactEncoded.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * as unpadded Base64 URL encoded strings.
+ */
+final class CompactEncoded {
+	private CompactEncoded() {
+	}
+
+	/**
+	 * Iterates over segments in a JSON compact serialized structure.
+	 * 
+	 * @param data compact serialize data
```
[the remark this answers is not in the document]
Comment (mwfyffeiu) [reply crypt/impl/src/main/java/iu/crypt/CompactEncoded.java]: Corrected with latest comment

File: crypt/impl/src/main/java/iu/crypt/CryptJsonAdapters.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	private CryptJsonAdapters() {
+	}
+
+	/**
+	 * JSON type adapter for byte[] using Base64URL encoding with no padding.
+	 */
+	public static final IuJsonAdapter<byte[]> B64URL = IuJsonAdapter.text(IuText::base64Url, IuText::base64Url);
+
+	/**
+	 * Singleton.
```
[the remark this answers is not in the document]
Comment (mwfyffeiu) [reply crypt/impl/src/main/java/iu/crypt/CryptJsonAdapters.java]: Resolved, with some additional context, by a8773ebf632e624489c971836185c69fdb6ddf4e

File: crypt/impl/src/test/java/iu/crypt/WebKeyTest.java [hunk 4]
```diff
@@ -849,25 +860,25 @@ public void testWellKnownWithOps() {
...
+//				key.getCertificateChain()[0].getIssuerX500Principal().getName());
+//
+//		key = WebKey.pem(CERT_TEXT);
+//		assertNull(key.getPrivateKey());
+//		assertNotNull(key.getPublicKey());
+//		assertNotNull(key.getCertificateChain());
+//		assertEquals(1, key.getCertificateChain().length);
+//		assertEquals("CN=iu-java-crypt-test,OU=STARCH,O=Indiana University,L=Bloomington,ST=Indiana,C=US",
+//				key.getCertificateChain()[0].getIssuerX500Principal().getName());
+//	}
```
[the remark this answers is not in the document]
Comment (mwfyffeiu) [reply crypt/impl/src/test/java/iu/crypt/WebKeyTest.java]: Not needed, removed by a8773ebf632e624489c971836185c69fdb6ddf4e

**Review by james-anderson-iu — APPROVED** [review]

**Comment by mwfyffeiu** [comment]
> Do we have any cases when converting between types with `from` and `to`, such as in `CryptJsonAdapters`, where it would be necessary to test both `from(to())` and `to(from())`?

I believe from(to()) covers adapter behavior, but to/from tests that verify interoperability with crypto tokens etc generated from other libraries/external sources are valuable to have. I added a few examples that prove JWT interactions with Jose4J, and there are more similar tests we can produce to more effectively cover exchanging signed and encrypted data with other systems.

**Pull request merged**